### PR TITLE
[ADD] server: optional timeout on blocking socket operations

### PR DESCRIPTION
**Pull request by sle-odoo**
workaround issue #2114 (explanation [here](https://github.com/odoo/odoo/issues/2114#issuecomment-102865136))

### CODE REVIEW

**Comment by odony** [comment]
I'm not sure about the default... shouldn't we put something large like 1s as a sensible default, and allow setting the config to 0 for no timeout. If 1 second turns out to be too small we can easily increase it.
This should help prevent accidental DOS, and kind of fix #2114 without manual config.
@antonylesuisse what do you think?

**Comment by antonylesuisse** [comment]
I would set more that 1s, something like 5s

File: openerp/tools/config.py [hunk 1]
```diff
@@ -281,6 +281,9 @@ def __init__(self, fname=None):
             group.add_option("--limit-request", dest="limit_request", my_default=8192,
                              help="Maximum number of request to be processed per worker (default 8192).",
                              type="int")
+            group.add_option("--limit-socket-timeout", dest="limit_socket_timeout", my_default=-1.0,
+                             help="Maximum time in second to spend in blocking socket operations (default: disabled).",
```
Comment (odony) [inline openerp/tools/config.py]: - second**s**

- Also it would be useful to mention that this setting only applies with workers>0, and only for HTTP workers

**Comment by sle-odoo** [comment]
@odony actually, setting a timeout of 0 second does more than disabling the timeout: it also [set the socket in non-blocking mode](https://docs.python.org/2/library/socket.html#socket.socket.settimeout). I suppose this information would also be useful in the the help :)

**Comment by odony** [comment]
@sle-odoo what I meant is that people would be able to pass `--limit-socket-timeout=0` and it would do the same as your current default `-1` value, i.e. not set the timeout at all (not set it to 0 ;-))

**Comment by flotho** [comment]
still happen on v10

**Comment by invitu** [comment]
We have a lot of socket timeouts in environment with (`apache reverse proxy -- firewall -- odoo server`)
The` --limit-socket-timeout` parameter would be useful for us too

**Comment by flotho** [comment]
@odony , any chance to see this PR merged ?

**Comment by odony** [comment]
@flotho This PR was written before we added a default socket timeout for 8.0+, about a year ago, in 8245c1d1d87b443701b161d8d4a42df2b4d13aec (hence the conflict).
Adding a config option would only give you control over the actual timeout, but the problem itself is already fixed if your code is up-to-date.

The PR can be merged in `master` after fixing it wrt to the review comments, and using a default timeout of 0.5, which has been battle-tested during 1y+

**Comment by invitu** [comment]
FYI, with Odoo v10c, workers = 7, firewall between reverse proxy and odoo server, we need to increase timeout to 2.0s for other. BTW, maybe we have some latency on our firewall
Here is a patch if it can help : 
`diff --git a/odoo/service/server.py b/odoo/service/server.py
index 1a9ec15..7e90b16 100644
--- a/odoo/service/server.py
+++ b/odoo/service/server.py
@@ -748,7 +748,8 @@ class WorkerHTTP(Worker):
     """ HTTP Request workers """
     def process_request(self, client, addr):
         client.setblocking(1)
-        client.settimeout(0.5)
+        if config['limit_socket_timeout'] != -1.0:
+            client.settimeout(config['limit_socket_timeout'])
         client.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
         # Prevent fd inherientence close_on_exec
         flags = fcntl.fcntl(client, fcntl.F_GETFD) | fcntl.FD_CLOEXEC
diff --git a/odoo/tools/config.py b/odoo/tools/config.py
index 5056ae7..55ee38a 100644
--- a/odoo/tools/config.py
+++ b/odoo/tools/config.py
@@ -284,6 +284,9 @@ class configmanager(object):
             group.add_option("--limit-request", dest="limit_request", my_default=8192,
                              help="Maximum number of request to be processed per worker (default 8192).",
                              type="int")
+            group.add_option("--limit-socket-timeout", dest="limit_socket_timeout", my_default=-1.0,
+                             help="Maximum time in seconds to spend in blocking socket operations (default: disabled). Only applies to HTTP workers with workers>0",
+                             type="float")
             parser.add_option_group(group)
 
         # Copy all optparse options (i.e. MyOption) into self.options.
@@ -415,7 +418,8 @@ class configmanager(object):
         posix_keys = [
             'workers',
             'limit_memory_hard', 'limit_memory_soft',
-            'limit_time_cpu', 'limit_time_real', 'limit_request', 'limit_time_real_cron'
+            'limit_time_cpu', 'limit_time_real', 'limit_request', 'limit_time_real_cron',
+            'limit_socket_timeout'
         ]
 
         if os.name == 'posix':
`

**Comment by pedrobaeza** [comment]
We have been experimenting also troubles with this configuration behind our Docker infrastructure with socket proxy + Traefik (also tried with haproxy, Nginx and Apache).

cc @Yajo

**Comment by odony** [comment]
If .5s is too short I see no problem to increase it to 1 or 2s in stable versions, it will simply delay the recovery in case of a worker lock due to wkhtmltopdf. I'm surprised that we've heard so little about it, though. We had a discussion on commit 8245c1d1d87b443701b161d8d4a42df2b4d13aec in August 2016 which did not yield any definitive conclusion, and that's all I've seen about it until now, I think.

Note that nowadays network latencies > 500ms are rather uncommon, even on very remote places (satellite uplinks would be slightly above it). So triggering the timeout should still be fairly difficult: you would need to hold a TCP socket open for 500ms without sending a single byte.

@invitu @pedrobaeza do you have more specific data regarding the connection timings? For what kind of requests do you see it (@invitu mentioned "other treatments")? Did you experiment with different timeout values? Did you analyze the traffic in the different components to see where the time was being spent?

FWIW we currently use the 500ms timeout in production even with multiple levels of load-balancing/proxying/firewall running on different physical machines.

**Comment by elicoidal** [comment]
> Note that nowadays network latencies > 500ms are rather uncommon, even on very remote places (satellite uplinks would be slightly above it). 

From China, network is pretty slow but still half the 500 ms

PING odoo.com (178.33.40.43) 56(84) bytes of data.
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=1 ttl=47 time=231 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=2 ttl=47 time=231 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=3 ttl=47 time=232 ms

**Comment by invitu** [comment]
@odony Here are some cases I picked up from our logs (most of them are web*.js): 
```
Jun 19 17:29:16 odoo journal: Odoo Server 10.0:db1:INFO:werkzeug:172.22.230.214 - - [20/Jun/2017 03:29:16] "GET /web/content/256-9982b33/web_editor.assets_editor.js HTTP/1.1" 200 -
Jun 19 17:29:17 odoo odoo-server_10.0: 172.22.230.214 - - [2017-06-20 03:29:17] "POST /longpolling/poll HTTP/1.1" 200 286 50.037857
Jun 19 17:29:18 odoo journal: Odoo Server 10.0:db1:ERROR:odoo.service.server:Worker (20434) Exception occured, exiting...#012Traceback (most recent call last):#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 737, in run#012    self.process_work()#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 770, in process_work#012    self.process_request(client, addr)#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 761, in process_request#012    self.server.process_request(client, addr)#012  File "/usr/lib64/python2.7/SocketServer.py", line 321, in process_request#012    self.finish_request(request, client_address)#012  File "/usr/lib64/python2.7/SocketServer.py", line 334, in finish_request#012    self.RequestHandlerClass(request, client_address, self)#012  File "/usr/lib64/python2.7/SocketServer.py", line 651, in __init__#012    self.finish()#012  File "/usr/lib64/python2.7/SocketServer.py", line 710, in finish#012    self.wfile.close()#012  File "/usr/lib64/python2.7/socket.py", line 279, in close#012    self.flush()#012  File "/usr/lib64/python2.7/socket.py", line 303, in flush#012    self._sock.sendall(view[write_offset:write_offset+buffer_size])#012timeout: timed out


Jun 20 09:21:27 odoo journal: Odoo Server 10.0:db2:INFO:werkzeug:172.22.230.214 - - [20/Jun/2017 19:21:27] "GET /web/content/347-63b8340/web.assets_backend.js HTTP/1.1" 200 -
Jun 20 09:21:29 odoo journal: Odoo Server 10.0:db2:ERROR:odoo.service.server:Worker (22369) Exception occured, exiting...#012Traceback (most recent call last):#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 737, in run#012    self.process_work()#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 770, in process_work#012    self.process_request(client, addr)#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 761, in process_request#012    self.server.process_request(client, addr)#012  File "/usr/lib64/python2.7/SocketServer.py", line 321, in process_request#012    self.finish_request(request, client_address)#012  File "/usr/lib64/python2.7/SocketServer.py", line 334, in finish_request#012    self.RequestHandlerClass(request, client_address, self)#012  File "/usr/lib64/python2.7/SocketServer.py", line 651, in __init__#012    self.finish()#012  File "/usr/lib64/python2.7/SocketServer.py", line 710, in finish#012    self.wfile.close()#012  File "/usr/lib64/python2.7/socket.py", line 279, in close#012    self.flush()#012  File "/usr/lib64/python2.7/socket.py", line 303, in flush#012    self._sock.sendall(view[write_offset:write_offset+buffer_size])#012timeout: timed out

Jun 20 09:29:26 odoo journal: Odoo Server 10.0:db3:INFO:werkzeug:172.22.230.214 - - [20/Jun/2017 19:29:26] "POST /report/download HTTP/1.1" 200 -
Jun 20 09:29:26 odoo odoo-server_10.0: 172.22.230.214 - - [2017-06-20 19:29:26] "POST /longpolling/poll HTTP/1.1" 200 286 50.038101
Jun 20 09:29:28 odoo journal: Odoo Server 10.0:db3:ERROR:odoo.service.server:Worker (20537) Exception occured, exiting...#012Traceback (most recent call last):#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 737, in run#012    self.process_work()#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 770, in process_work#012    self.process_request(client, addr)#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 761, in process_request#012    self.server.process_request(client, addr)#012  File "/usr/lib64/python2.7/SocketServer.py", line 321, in process_request#012    self.finish_request(request, client_address)#012  File "/usr/lib64/python2.7/SocketServer.py", line 334, in finish_request#012    self.RequestHandlerClass(request, client_address, self)#012  File "/usr/lib64/python2.7/SocketServer.py", line 651, in __init__#012    self.finish()#012  File "/usr/lib64/python2.7/SocketServer.py", line 710, in finish#012    self.wfile.close()#012  File "/usr/lib64/python2.7/socket.py", line 279, in close#012    self.flush()#012  File "/usr/lib64/python2.7/socket.py", line 303, in flush#012    self._sock.sendall(view[write_offset:write_offset+buffer_size])#012timeout: timed out

Jun 20 12:53:21 odoo journal: Odoo Server 10.0:db4:INFO:werkzeug:172.22.230.214 - - [20/Jun/2017 22:53:21] "GET /web/content/661-85e56d5/web.assets_common.js HTTP/1.1" 200 -
Jun 20 12:53:24 odoo odoo-server_10.0: 172.22.230.214 - - [2017-06-20 22:53:24] "POST /longpolling/poll HTTP/1.1" 200 386 50.038443
Jun 20 12:53:24 odoo journal: Odoo Server 10.0:db4:ERROR:odoo.service.server:Worker (13356) Exception occured, exiting...#012Traceback (most recent call last):#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 737, in run#012    self.process_work()#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 770, in process_work#012    self.process_request(client, addr)#012  File "/var/opt/odoo/10.0/server_20170606/odoo/service/server.py", line 761, in process_request#012    self.server.process_request(client, addr)#012  File "/usr/lib64/python2.7/SocketServer.py", line 321, in process_request#012    self.finish_request(request, client_address)#012  File "/usr/lib64/python2.7/SocketServer.py", line 334, in finish_request#012    self.RequestHandlerClass(request, client_address, self)#012  File "/usr/lib64/python2.7/SocketServer.py", line 651, in __init__#012    self.finish()#012  File "/usr/lib64/python2.7/SocketServer.py", line 710, in finish#012    self.wfile.close()#012  File "/usr/lib64/python2.7/socket.py", line 279, in close#012    self.flush()#012  File "/usr/lib64/python2.7/socket.py", line 303, in flush#012    self._sock.sendall(view[write_offset:write_offset+buffer_size])#012timeout: timed out
```
With 500ms, we had around 100 errors / day
With 1.5s we still had a few errors
With 2.0s, we did not have errors for last 2 days

Here in Tahiti, network can be slow/bad quality... but I guess it can be worse in a lot of countries, so it would be could to put this setup as a config parameter instead as hardcoding it

**Comment by invitu** [comment]
@elicoidal from Tahiti : 
PING odoo.com (178.33.40.43) 56(84) bytes of data.
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=1 ttl=41 time=304 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=2 ttl=41 time=307 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=3 ttl=41 time=329 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=4 ttl=41 time=352 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=5 ttl=41 time=375 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=6 ttl=41 time=398 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=7 ttl=41 time=318 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=8 ttl=41 time=341 ms
64 bytes from ip43.ip-178-33-40.eu (178.33.40.43): icmp_seq=9 ttl=41 time=364 ms

**Comment by pedrobaeza** [comment]
You can easily reproduce it in any place using Developer tools from the browser and activating a bandwith limit in them.

**Comment by elicoidal** [comment]
@invitu ... and I thought I was unlucky :stuck_out_tongue_closed_eyes:

**Comment by antonylesuisse** [comment]
So 2s default ?

**Comment by odony** [comment]
The default timeout has been increased to 2s via #17998, as of 8.0 (to be forward-ported soon)

**Comment by odony** [comment]
After changing the default to 2s, do you think it is still useful to make it a config option? We have so many options already...

**Comment by invitu** [comment]
As we still had issues with 2 s, we went to 3 s with our prod instance 2 days ago... :-(

**Comment by Yajo** [comment]
We have a customer with specially low bandwidth. I had to raise it to 5s for some of his corner cases, but didn't want to put 5s in #17998 because it seems too much for general use. We'd appreciate much to have this setting, if you don't mind 😊

**Comment by ap-wtioit** [comment]
We had a problem with low bandwith as well (70Mbps) for a database of 1.6GB in /web/database/manager it constantly hits the 2s limit

**Comment by wtaferner** [comment]
@odony 
Why was this useful option not merged in the end? Do we have an explicit contradiction to make this configurable? I guess the reason for the low timeout was to fast-recycle HTTPWorkers if they become unresponsive, right?

**Comment by odony** [comment]
@wtaferner the arguments are in the discussion above:
- the timeout was a solution for #2114 and it appears to be working well
- after bumping the timeout to 2s we've had close to zero tickets about this, so the default is good
- as usual, we want to avoid adding server options "just in case"

Now please notice how many come here in search of a solution for a *low bandwidth* situation, even though the timeout is related to *latency*, not bandwidth.

Very high network latencies and packet loss are a symptom of a congested outbound network path: buffers are filling up and packets are delayed or dropped. In this situation the latency can go arbitrarily high. People who have used bittorrent on a residential connection are probably familiar with this.

In the Odoo world that could typically occur when downloading a large file (e.g. db dump) from a server with a bandwidth that is much lower than the bandwidth of the receiving side.

A better solution to such problems is to introduce QoS on the network to avoid saturating the whole network with a single TCP stream, avoiding congestion and packet loss altogether, making the whole transfer faster. If there is no network equipment capable of QoS, an alternative is to upload the file with a program that supports bandwidth capping, such as `scp -l`, and put the cap a little bit below the BW limit.

That seems much better in the long run than bumping the socket timeout to a random large value, trying to ignore the congestion yet fueling it, and impacting HTTPWorker lock detection permanently as a side-effect.

In other words, this setting would only be a bad workaround for a bandwidth saturation problem.

PS: To anyone coming here in the future: please, please do not use the DB manager for anything else than local dev/testing. [It should always be disabled!](https://www.odoo.com/documentation/13.0/setup/deploy.html#database-manager-security) If you're trying to download a large DB directly from the Odoo DB manager you're doing something wrong.

**Comment by wtaferner** [comment]
@odony 
Thank you for explaining for everyone the issue for low bandwidth. Our use case is a different issue and I guess it is not yet supported by the HTTPWorker Implementation. Any plans to make a stable stream possible without bumping up the socket timeout limits?

**Comment by cvinh** [comment]
@odony for your information we had to increase to 20 seconds on our production (v12) instance ! because we have the timeout problem when downloading not so big file (3-4Mb) in attachment to a message.

**Comment by odony** [comment]
@cvinh thanks for the info! Is this a bandwidth saturation issue too? What's your server-side bandwidth and latency? Is it shared with other downloads? 

Have you considered using a QoS system? Even something as simple as a nginx rule with [`limit_rate`](http://nginx.org/en/docs/http/ngx_http_core_module.html#limit_rate) might be able to keep each connection under the saturation point, and make the whole system more responsive for all services, not just Odoo.

**Comment by cvinh** [comment]
No it's just that the bandwidth is 4Mbps and a 4MB file takes 10 seconds to download... we put 20 seconds just to secure the customer.

**Comment by cvinh** [comment]
cc @luceeasypme

**Comment by eLBati** [comment]
Timeout can now be set by env variable: https://github.com/odoo/odoo/commit/85fe44a7298ef9883160359814e74e39b7e10873
